### PR TITLE
Add simple LRU cache

### PR DESCRIPTION
**Pull request by corny**

### CODE REVIEW

File: lib/texd/attachment.rb [hunk 1]
```diff
@@ -202,9 +202,21 @@ def to_upload_io(full: false)
         }
       end
 
+      @cache = Cache.new(128)
```
Comment (dmke) [inline lib/texd/attachment.rb:205]: How do we make this number configurable?

File: lib/texd/cache.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    class Node
+      attr_reader :key
+      attr_accessor :value, :prev, :next
+
+      def initialize(key)
+        @key = key
+      end
+    end
+
+    attr_reader :list, :hash, :count, :capacity
```
Comment (dmke) [inline lib/texd/cache.rb:45]: Should `hash` be publicly accessible? AFAICT, its usefulness is limited to the test suite.

File: spec/lib/texd/cache_spec.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    cache.write(:four, "4")
+    expect(cache.count).to eq 3
+
+    expect(cache.hash.keys).to include(:two, :three, :four)
+
+    cache.read(:two)
+    cache.read(:four)
+    cache.write(:five, "5")
+    expect(cache.hash.keys).to include(:two, :four, :five)
```
Comment (dmke) [inline spec/lib/texd/cache_spec.rb:32]: Interestingly, `expect(cache.hash.keys).to match [:five, :two, :four]` fails.

I know that hashes keep the keys in insert order, but nevertheless this had surprised me.

File: lib/texd/cache.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      @list     = LinkedList.new
+      @hash     = {}
+      @mtx      = Mutex.new
+      @count    = 0
+      @capacity = capacity
+    end
+
+    def lookup(key)
+      read(key) || write(key, yield)
+    end
```
Comment (dmke) [inline lib/texd/cache.rb:57]: I've added `#lookup`, it makes using the cache a bit more idiomatic:

```rb
cache.lookup(:foo) { "expensive computation" }
```

I'd like a better name though, as it performs a "lookup or store" action. The syntax is inspired by Rails' fragment caching, where we can do

```erb
<% cache "key" do %>
  <p>HTML fragment</p>
<% end %>
```

Re-using the name "cache" feels wrong on the call site though:

```rb
cache.cache(:foo) { "expensive computation" }
```

Any opinions?

Comment (corny) [reply lib/texd/cache.rb:57]: Good idea!

File: lib/texd/cache.rb (hunk 2 above)
Comment (corny) [reply lib/texd/cache.rb]: No does not need to be publicly accessible, I removed it.

File: lib/texd/cache.rb (hunk 4 above)
Comment (dmke) [reply lib/texd/cache.rb]: We're going with `#fetch`.

**Pull request merged**